dx12: Implement root constants

msiglreith opened the pull request:
Implement root constants for the d3d12 backend following https://github.com/gfx-rs/gfx/issues/1631
Push constants will be split into disjunct ranges, which can be mapped to root constants in the root signature. On pipeline creation, the generated HLSL shaders from spirv-cross will be patched to match the root signature.

- [x] Split push constants into disjunct ranges
- [x] Update root signature for root constants support
- [x] Patch generated descriptor set table register spaces
- [x] Patch push constant translation in spirv-cross

~~This PR is still in an early state. Especially, doing the spirv-cross changes will take a bit of time.~~

Closes https://github.com/gfx-rs/gfx/issues/1625 https://github.com/gfx-rs/gfx/issues/1631

msiglreith commented:
PR is should be ready, just waiting on a new spirv-cross release atm (I just openend the PR).

kvark commented on src/backend/dx12/src/command.rs:
```diff
@@ -341,7 +345,15 @@ impl CommandBuffer {
+                    0,
+                )
+            },
+            |slot, gpu| unsafe { cmd_buffer.clone().SetGraphicsRootDescriptorTable(slot, gpu); },
```
nit: let's have this expression on multiple lines

kvark commented on src/backend/dx12/src/command.rs:
```diff
@@ -357,44 +369,75 @@ impl CommandBuffer {
+        let num_root_constant = pipeline.root_constants.len();
+        let mut cur_index = 0;
+        // TODO: opt: Only set dirty root constants?
+        for i in 0..num_root_constant {
```
nit: use `iter().enumerate()`

kvark commented on src/backend/dx12/src/command.rs:
```diff
@@ -357,44 +369,75 @@ impl CommandBuffer {
                         pipeline.sampler_start + offset as u64,
                     _ => {
-                        error!("Unexpected user data element in the root signature ({:?})", user_data.data[i]);
+                        error!("Unexpected user data element in the root signature ({:?})", user_data.data[table_index]);
```
nit: match with `other` instead of `_` and use it here

kvark commented on src/backend/dx12/src/command.rs:
```diff
@@ -357,44 +369,75 @@ impl CommandBuffer {
-                table_update(slot, gpu);
-                user_data.clear_dirty(table_index);
+                table_update(i as _, gpu);
+                user_data.clear_dirty(i);
```
should this use `table_index`?

kvark commented on src/backend/dx12/src/command.rs:
```diff
@@ -1412,11 +1467,22 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
+               root_constant.range.start < offset+num
+            {
+                let start = (root_constant.range.start-offset) as _;
+                let end = num.min(root_constant.range.end-offset) as _;
```
shouldn't this assert/warn if the end is beyond the available range?

kvark commented on src/backend/dx12/src/command.rs:
```diff
@@ -1425,7 +1491,18 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
+            {
+                let start = (root_constant.range.start-offset) as _;
+                let end = num.min(root_constant.range.end-offset) as _;
+                self.comp_pipeline
```
would be nice to re-use this code

kvark commented on src/backend/dx12/src/device.rs:
```diff
@@ -160,31 +163,66 @@ impl Device {
         compile_options.shader_model = shader_model;
         compile_options.vertex.invert_y = true;
 
+        let stage_flag = match stage {
```
we should probably implement `From<pso::Stage> for pso::ShaderStageFlags`, if it isn't already

kvark commented on src/backend/dx12/src/device.rs:
```diff
@@ -798,17 +840,60 @@ impl d::Device<B> for Device {
         // separated, so each set layout will actually occupy up to 2 entries!
+        //
+        // Root signature layout:
+        //     Root Constants: Register: Offest/4, Space: 0
```
👍

kvark commented on src/backend/dx12/src/device.rs:
```diff
@@ -798,17 +840,60 @@ impl d::Device<B> for Device {
+            .map(|constant| {
+                RootConstant {
+                    stages: constant.stages,
+                    range: (constant.range.start/4)..(constant.range.end/4),
```
any checks needed to see if remainder by 4 is not zero?

msiglreith replied:
Hm, I should just adjust the API to interpret the range as multiples of u32.

kvark commented on src/backend/dx12/src/device.rs:
```diff
@@ -798,17 +840,60 @@ impl d::Device<B> for Device {
+
+            *unsafe{ param.Constants_mut() } = winapi::D3D12_ROOT_CONSTANTS {
+                ShaderRegister: root_constant.range.start as _,
+                RegisterSpace: 0, // root constants will always be in space0.
```
nit: could make it a constant with a nice name at the top of the module

kvark commented on src/backend/dx12/src/device.rs:
```diff
@@ -1143,31 +1242,7 @@ impl d::Device<B> for Device {
-            shader_map.insert(entry_point.name, shader_blob);
-        }
-        Ok(n::ShaderModule::Compiled(shader_map))
+        Ok(n::ShaderModule::Spirv(raw_data.into()))
```
can we detect here if any push constants are present and go to the `Compiled` fast path?

msiglreith replied:
Hm requires two things:
- Checking if it's allowed to not declare constants buffers from the root signature in the shaders
- Adjust register spaces of descriptor tables to enforce that space0 will always be reserved for root constants even if the pipeline layout doesn't have push constants.

kvark commented on src/backend/dx12/src/root_constants.rs:
```diff
@@ -0,0 +1,302 @@
+    // Divide a root constant into two separate ranges depending on the overlap
+    // with another root constant.
+    fn divide(self, other: &RootConstant) -> (RootConstant, RootConstant) {
+        let left = RootConstant {
```
any checks for ordering? e.g. `self.range.start <= other.range.start` and such

kvark commented on src/backend/dx12/src/root_constants.rs:
```diff
@@ -0,0 +1,302 @@
+
+impl PartialOrd for RootConstant {
+    fn partial_cmp(&self, other: &RootConstant) -> Option<Ordering> {
+        Some(self.range.start.cmp(&other.range.start)
```
technically if we reorder `stages` to go last, we could derive the `Ord` and `PartialOrd`, right?

msiglreith replied:
Not sure what you mean.

kvark replied:
Nvm, I see that `Range` doesn't implement `PartialOrd`. Was thinking about putting `#[derive()]` instead of manual implementation, but it's not possible.

kvark commented on src/backend/dx12/src/root_constants.rs:
```diff
@@ -0,0 +1,302 @@
+}
+
+impl<'a> From<&'a (pso::ShaderStageFlags, Range<u32>)> for RootConstant {
+    fn from(&(stages, ref range): &'a (pso::ShaderStageFlags, Range<u32>)) -> Self {
```
where is this one needed?

msiglreith replied:
It's used in `into_vec`.

kvark replied:
Let's move this code into the `into_vec` implementation then. I don't think this `From<>` is a good thing to have here.

kvark commented on src/backend/dx12/src/root_constants.rs:
```diff
@@ -0,0 +1,302 @@
+    // Frontier of unexplored root constant ranges, sorted descending
+    // (less element shifting for Vec) regarding to the start of ranges.
+    let mut ranges = into_vec(ranges);
+    ranges.sort_by(|a, b| a.cmp(b).reverse());
```
wouldn't this just be `b.cmp(a)`?

msiglreith replied:
r-right... :facepalm:

kvark commented on src/backend/dx12/src/root_constants.rs:
```diff
@@ -0,0 +1,302 @@
+        } else {
+            disjunct.push(cur);
+        }
+        ranges.sort_by(|a, b| a.cmp(b).reverse());
```
similarly here

kvark commented on src/backend/dx12/src/root_constants.rs:
```diff
@@ -0,0 +1,302 @@
+    // Storing resulting disjunct root constant ranges.
+    let mut disjunct = Vec::with_capacity(ranges.len());
+
+    while !ranges.is_empty() {
```
could start here right away with `while let Some(cur) = ranges.pop()`

kvark commented on src/backend/dx12/src/root_constants.rs:
```diff
@@ -0,0 +1,302 @@
+                range: 3..4,
+            },
+        ];
+        assert_eq!(reference, split(ranges));
```
neat!

msiglreith commented:
Thanks for the extensive review, I think I addressed all points.
Just noticed that push constant updates are bugged due to an issue in `ash`, but that's slightly out-of-scope for this PR. Requires a version bump in the future.

kvark commented:
I'm planning on getting the tests for this.
Thanks for addressing the issues!
bors r+

> On Dec 2, 2017, at 11:22, msiglreith <notifications@github.com> wrote:
> 
> Thanks for the extensive review, I think I addressed all points.
> Just noticed that push constant updates are bugged due to an issue in ash, but that's slightly out-of-scope for this PR. Requires a version bump in the future.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
